fontbe/src/glyphs.rs

```diff
@@ -258,14 +265,49 @@ impl Work<Context, Error> for GlyphWork {
                 };
                 context.set_glyph(name.clone(), base_glyph.clone().into());
 
-                (name, point_seqs_for_simple_glyph(ir_glyph, instances))
+                let mut contour_end = 0;
+                let mut contour_ends = Vec::with_capacity(base_glyph.contours().len());
+                for contour in base_glyph.contours() {
+                    contour_end += contour.len();
+                    contour_ends.push(contour_end);
+                }
+                (
+                    name,
+                    point_seqs_for_simple_glyph(ir_glyph, instances),
+                    contour_ends,
+                )
             }
         };
 
+        let coords = point_seqs.get(default_location).ok_or_else(|| {
+            Error::GlyphError(ir_glyph.name.clone(), GlyphProblem::MissingDefault)
+        })?;
+
+        // TODO real values
+        let tolerance = 0.1;
+
         // Contour (aka Simple) and Composite both need gvar
         let deltas = var_model
             .deltas(&point_seqs)
-            .map_err(|e| Error::GlyphDeltaError(self.glyph_name.clone(), e))?;
+            .map_err(|e| Error::GlyphDeltaError(self.glyph_name.clone(), e))?
+            .into_iter()
+            .map(|(region, deltas)| {
+                // Spec: inferring of deltas for un-referenced points applies only to simple glyphs, not to composite glyphs.
+                if should_iup {
+                    // Doing IUP optimization here conveniently means it threads per-glyph
+                    if log_enabled!(log::Level::Trace) {
+                        // I like the point string better than the vec2
+                        let deltas = deltas.iter().map(|d| d.to_point()).collect::<Vec<_>>();
+                        trace!("IUP '{}', tolerance {tolerance}\n  {} contour ends {contour_ends:?}\n  {} deltas {deltas:?}\n  {} coords {coords:?}", self.glyph_name, contour_ends.len(), deltas.len(), coords.len());
+                    }
+                    iup_delta_optimize(deltas, coords.clone(), tolerance, &contour_ends)
+                        .map(|iup_deltas| (region.clone(), iup_deltas))
+                } else {
+                    Ok((region, deltas.into_iter().map(Some).collect()))
+                }
+            })
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(|e| Error::IupError(ir_glyph.name.clone(), e))?;
 
         context.set_gvar_fragment(name, GvarFragment { deltas });
```

Comment:
I think you should check if all deltas are None (because they are all implied, meaning that the deltas were originally all no-op zeros), and if so we could just skip the GvarFragment since it doesn't do anything.. Unless something else down the pipeline is already taking care of that?

At least for simple glyphs.. remember we do _not_ want to drop these all-zero gvar tuples for the composite glyphs to match the fonttools workaround at https://github.com/googlefonts/fontmake-rs/issues/283

Reply:
I'm not doing iup for composites at all which I think should suffice?

Good point wrt all None.

Reply:
I'm merrily producing empty deltas. I think I should be filtering them out indeed.

Reply:
Fixed